### DISCUSSION TITLE
Why does WDS recompile immediately after a successful startup

### DISCUSSION THREAD

**Discussion by darkyzhou** [webpack/webpack#16228] [Q&A] [answered]
Recently I am learning internal details of Webpack.

For some complex frontend projects, I always encounter a strange case:

After I run webpack-dev-server and it compiles successfully, I will get: `Build completed in XX.XXs`

Immediately, webpack-dev-server starts compiling again and seems to only compile a small number of modules. Soon I get another `Build completed in YY.YYs`

After setting some breakpoints, I notice that the modules(mainly `NormalModule`s in my case) compiled in the secend run were also compiled in the first one, but with different `hash`.

So my questions are:
* Why does webpack-dev-server need a second run of compilation?
* Is it because some modules(or their dependencies) have changed? But how can they be changed during the first run?
* Is it possible to avoid the second run?

I might miss some details, feel free to ask me for them.

Any help will be appreciated, thanks!

**Comment by alexander-akait** [comment, marked as answer]
You should not have two compilation, can you provide your configuration?

**Reply by darkyzhou** [reply]
> You should not have two compilation, can you provide your configuration?

Thank you. But I am afraid I can't because it is from a project of the company I am working at.

Can you share some ideas about what could potentially be the reasons? Or I will spend some time trying if I can reproduce the problem.

BTW, the project is still using webpack 4, but I remember encountering the same problem in webpack 5. So I thought it is somehow a feature before.

**Reply by alexander-akait** [reply]
Do you have a copy-webpack-plugin or something that can copy/modify the file once it's run?

**Reply by darkyzhou** [reply]
> Do you have a copy-webpack-plugin or something that can copy/modify the file once it's run?

Thank you! I think I have found the reason. Let me have some time doing some research about it.

**Reply by darkyzhou** [reply]
Sorry for being late. There were two reasons:
* [antd-dayjs-webpack-plugin](https://github.com/ant-design/antd-dayjs-webpack-plugin). It will construct another `Compilation` after the first run due to its way of implementation, thus causing the second run.
* [fork-ts-checker-webpack-plugin](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin). Using tsc to perform type checking, some symbolic files of the project will be modified somehow, causing webpack's `Watcher` to be notified, thus leading to the second run.
